### DISCUSSION TITLE
Dual Power Ocean Configuration

### DISCUSSION THREAD

**Discussion by AndyBowden** [niltrip/powerocean#10] [General]
I had a Power Ocean system installed yesterday.
I’ve already installed and set up this HA integration without difficulty.
My thanks go to niltrip and all other contributors.
My Power Ocean system comprises two DC single phase inverters each with a pair of 5kWh batteries. One of the inverters is master and the other is slave.
I set up the integration by entering the serial number of the master. I got a total of 45 entities associated with the integration.
Some of these are live with an update about every 1 minute. Many of the entities seem to relate just to the master inverter and its 2 batteries. For example, sensor.powerocean_bpremainwatth peaks at 10,000Wh whilst the total battery capacity is 20kWh and was full overnight.
I created a second device for the integration using the serial number of the slave inverter. The device was created, but it has no entities associated with it.
I presume that my dual inverter installation is unusual and no other users of this integration have a similar configuration yet. I further presume that the ECOFlow API is not making data for the second inverter/battery pair available in a manner suitable for this integration.
If there’s anyone with experience of this Power Ocean configuration, I’m keen to share knowledge.
On a second point, so far as I can tell this integration provides for monitoring of the Power Ocean system and does not provide any control, of charging periods or levels, for example.

**Comment by niltrip** [comment]
Hello, yes your constellation seems to be more special. I think that the response I get from the web portal for my system will look different than yours. You can try generating a response from your system and compare it with mine. Mine is at documentation/response_modified_5_1_16_11.json. If you know the differences, you can definitely do something. 

Generating a response is easy: activate debug and then reload the integration. The log file should then contain the complete answer. It should start with 

`[custom_components.powerocean] {'code': '0', 'message': 'Success',...`

Copy the string ( within the curly brackets {} ) into an editor and format it a little and save it as json. I like to use notepad++ and a plugin “JsonTools”.


Remember to remove the serial numbers and other data related to you, then you can provide the file.

**Reply by AndyBowden** [reply]
Thanks for your suggestion. My apologies for the slow response.
My system has now been fully commissioned after some issues.
I've been able to generate a response and it differs from yours somewhat.
I'm a Mac user and I've not managed to find a json conversion equivalent to your suggested notepad++ plus JsonTools. I attach a raw text file which may be of interest.
The file does include the serial numbers of both inverters and all 4 batteries.
There are 44 blocks commencing JTS with 20 distinct JTS strings.
[EcoFlow Report San3.txt](https://github.com/user-attachments/files/17701924/EcoFlow.Report.San3.txt)

**Comment by niltrip** [comment]
Hello, yes the response from the server is twice as large as that of a normal system. There is a special area for parallel operation. I will take a closer look at it over the weekend.

Can you try to generate a response with the slave serial? Perhaps it is possible to create two devices. Otherwise I will have to look at how this constellation can be appropriately mapped.

[EcoFlow.Report.San3.json](https://github.com/user-attachments/files/17738863/EcoFlow.Report.San3.json)

**Reply by AndyBowden** [reply]
Thanks for your response and the json file version.
Early on I did create a second device using the serial number of the slave inverter. The integration did accept that and HA showed a second device, but it had no entities.
The technicians who commissioned the system believed that all communications from the system were via the master. The slave's IOT dongle is not live.
Certainly, the ECOFlow iPhone app refuses to link to the slave inverter directly, but does show the full system.

**Reply by AndyBowden** [reply]
I’ve had time to examine more debug data logged by the integration for my dual inverter Power Ocean system.
[Friday String.txt](https://github.com/user-attachments/files/17871334/Friday.String.txt)
The data has two big sections headed quota and parallel respectively.
The parallel section has two parts one relating to the master inverter and the other the slave.
The information in the parallel master section largely duplicates the non-void parts of the quota section.
Comparison of the master and slave sections indicates that only the JTS1_EMS_CHANGE_REPORT, JTS1_EMS_HEARTBEAT and JTS1_ENERGY_STREAM_REPORT blocks contain data which differs between master and slave.
I attach a table which summarises my findings.
I hope this information is of use.
[Structure ComparisonV2.pdf](https://github.com/user-attachments/files/17871062/Structure.ComparisonV2.pdf)

**Reply by AndyBowden** [reply]
I've made good progress producing a version of ecoflow.py which sets up the sensors of both the master and slave inverters in my dual inverter Power Ocean system. It results in 115 sensor entities.
I've got some tidying up and further testing to do before I could share it

**Reply by niltrip** [reply]
Nice to hear that you started yourself. Unfortunately, I haven't found the time to look into your problem in more detail. Do you want to integrate it into this repo?

**Reply by AndyBowden** [reply]
It will be sensible for me to do so once I've investigated things a bit further. I've had some time available in the last couple of weeks, but I'm quite busy now and I have some time away soon so completion is likely to be in January.

**Reply by AndyBowden** [reply]
A question for you niltrip.
It seems to me that the naming of sensor entities by the function _get_sensors_battery_ is not consistent, with bpTemp being handled differently from the other sensors.
Did you have a reason for this?
I appreciate that the objective is to calculate the average of cell temperatures, but the naming convention seems to differ.
Thanks Andy

**Reply by niltrip** [reply]
> A question for you niltrip. It seems to me that the naming of sensor entities by the function _get_sensors_battery_ is not consistent, with bpTemp being handled differently from the other sensors. Did you have a reason for this? I appreciate that the objective is to calculate the average of cell temperatures, but the naming convention seems to differ. Thanks Andy

Hello, I've been very busy the last few days. One more week and it'll be the end of the year. I hope I'll find time then.
I can't see any difference in the names at first glance. But yes, I've tried to stay consistent.

**Reply by AndyBowden** [reply]
Sorry, I've been quiet for a while. The apparent inconsistency of sensor naming was my imagination.
I've produced a version of ecoflow.py which gives sensible answers for my dual inverter system and, I believe, will still work for single inverter systems.
I'll post it as a branch in your repository in January, niltrip. It will need testing for the single inverter case and for ackmachine's which only has batteries connected to the first inverter.

**Comment by Ackmachine** [comment]
Hey, I believe I have a similar configuration to AndyBowden. Here is my .json file from a log a few days ago. Maybe it can help. Thanks for your work here!

[Ecoflow2.json](https://github.com/user-attachments/files/18036419/Ecoflow2.json)

**Reply by AndyBowden** [reply]
That's useful thanks.
A quick inspection shows your file has one significant difference from mine. The JTS1_BP_STA_REPORT segment for the second inverter is very short containing just 'update time'.
Do you have batteries connected to the second inverter?

**Reply by Ackmachine** [reply]
Nope, no batteries on the second inverter. just the two 5kwh batteries on the first inverter.

**Comment by AndyBowden** [comment]
I have now created a branch of this repository containing my modified version of ecoflow.py which handles my dual inverter Power Ocean  installation.
My version should still function with single inverter installations, but I am unable to test it on one.
Documentation also contains a list of the 115 entities created by the modified integration  for my  installation.
I've raised a Pull Request.
Thanks
Andy

**Reply by Ackmachine** [reply]
I’ve tested your branch on my system, and everything worked without any issues until today. But that’s another topic (see: Issue #16).

The only thing I noticed is that I believe my master and slave are inverted. The battery shows up as a slave, but it’s actually installed on the master inverter.

Other than that, everything worked flawlessly, and all sensors were functioning correctly.

**Reply by AndyBowden** [reply]
Thanks.
I may have got master and slave reversed somewhere.
I'll investigate if and when the bigger issue #16 is resolved

**Reply by AndyBowden** [reply]
The header changes by Fabian Martin in https://github.com/niltrip/powerocean/issues/16#issuecomment-2575439350
have got things working again for me.
However, for reasons I am yet to determine, my integration with those header changes incorporated no longer captures data on the slave inverter of my dual inverter installation.

**Reply by Ackmachine** [reply]
Weird, for me it works and I have data on the master and slave. I am using your latest commit b6dc761.

**Reply by AndyBowden** [reply]
Very weird indeed, the json structure I am getting today is very different from yesterday's. If you are able would you let me have a copy of the json you are getting today, please.

Also what is the exact form of the header definition you are using at line 118?

Thanks

**Reply by AndyBowden** [reply]
I think I've explained my problem. Somehow, my repository had reverted to a November version of ecoflow.py.
Everything seems right now.
I'll look into the master / slave confusion on Thursday. Busy tomorrow.

**Reply by Ackmachine** [reply]
Ah ok. I am pretty busy myself at the moment.

**Reply by AndyBowden** [reply]
I have produced a new version of ecoflow.py which can handle master data segment before slave within the parallel segment or the other way round.
In my case, slave precedes master and I wrongly assumed that would always be the case.

In fact, the order of the segments seems to by the alphanumeric order of serial numbers. My master serial no is xxx0104 and slave xxx0042 so slave precedes master in my data.

Hopefully this version will work fine for your configuration Ackmachine
